REGRESSION(262863@main) [Cocoa] video.buffered ranges are always empty; breaks Shaka Player

jernoble opened the pull request:
#### 74481ec4b07546952d73c80f973f5f2f08a30841
<pre>
REGRESSION(262863@main) [Cocoa] video.buffered ranges are always empty; breaks Shaka Player
<a href="https://bugs.webkit.org/show_bug.cgi?id=257400">https://bugs.webkit.org/show_bug.cgi?id=257400</a>
rdar://108202807

Reviewed by Eric Carlson and Jean-Yves Avenard.

In 262863@main, a change was made to no longer create TimeRanges and PlatformTimeRanges objects
in the course of returning bufferedRanges(). However, in so doing, development code was left in
the patch which would always return empty ranges from MediaPlayerPrivateAVFoundation::buffered().
Additionally, existing code that cleared buffered ranges from RemoteMediaPlayerState was changed
in such a ways as subsequent updates would always get empty ranges. The test which would have
caught this breakage has been disabled since 2013 (143932@main).

Re-enable the media/video-buffered.html test.

Don&apos;t re-create m_buffered every time a client calls platformBufferedTimeRanges(). Instead, re-create
m_buffered only when loadedTimeRangesDidChange(). Send an optional&lt;&gt; buffered range across the GPU
process boundary in RemoteMediaPlayerState, and only update MediaplayerPrivateRemote&apos;s buffered
value when the state has a valid buffered range.

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.cpp:
(WebCore::MediaPlayerPrivateAVFoundation::buffered const):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::cancelLoad):
(WebCore::MediaPlayerPrivateAVFoundationObjC::platformBufferedTimeRanges const):
(WebCore::MediaPlayerPrivateAVFoundationObjC::platformMaxTimeLoaded const):
(WebCore::MediaPlayerPrivateAVFoundationObjC::loadedTimeRangesDidChange):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::sendCachedState):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::updateCachedState):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerState.h:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerState.serialization.in:

Canonical link: <a href="https://commits.webkit.org/264629@main">https://commits.webkit.org/264629@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ba6807612b2d54cf2b2cec950b62d0c94311bd60

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8103 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8389 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8604 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9758 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8194 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8109 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10376 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8300 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11055 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8247 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9326 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7375 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9879 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6627 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7417 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14986 "131 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7749 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7544 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10894 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8011 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6515 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7319 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1970 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11527 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7754 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
```diff
@@ -1720,17 +1720,9 @@ static URL conformFragmentIdentifierForURL(const URL& url)
 {
-    using namespace PAL; // For CMTIMERANGE_IS_EMPTY.
-
     if (!m_avPlayerItem)
         return PlatformTimeRanges::emptyRanges();
```
Won't `m_buffered` be empty if `m_avPlayerItem` is NULL?

jernoble committed:
REGRESSION(262863@main) [Cocoa] video.buffered ranges are always empty; breaks Shaka Player
https://bugs.webkit.org/show_bug.cgi?id=257400
rdar://108202807

Reviewed by Eric Carlson and Jean-Yves Avenard.

In 262863@main, a change was made to no longer create TimeRanges and PlatformTimeRanges objects
in the course of returning bufferedRanges(). However, in so doing, development code was left in
the patch which would always return empty ranges from MediaPlayerPrivateAVFoundation::buffered().
Additionally, existing code that cleared buffered ranges from RemoteMediaPlayerState was changed
in such a ways as subsequent updates would always get empty ranges. The test which would have
caught this breakage has been disabled since 2013 (143932@main).

Re-enable the media/video-buffered.html test.

Don't re-create m_buffered every time a client calls platformBufferedTimeRanges(). Instead, re-create
m_buffered only when loadedTimeRangesDidChange(). Send an optional<> buffered range across the GPU
process boundary in RemoteMediaPlayerState, and only update MediaplayerPrivateRemote's buffered
value when the state has a valid buffered range.

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.cpp:
(WebCore::MediaPlayerPrivateAVFoundation::buffered const):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::cancelLoad):
(WebCore::MediaPlayerPrivateAVFoundationObjC::platformBufferedTimeRanges const):
(WebCore::MediaPlayerPrivateAVFoundationObjC::platformMaxTimeLoaded const):
(WebCore::MediaPlayerPrivateAVFoundationObjC::loadedTimeRangesDidChange):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::sendCachedState):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::updateCachedState):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerState.h:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerState.serialization.in:

Canonical link: https://commits.webkit.org/264629@main
